[ doc, build_doc ] Optionally bind to global wan when previewing

martin-lueker opened the pull request:
This commit adds an additional --bind-wan argument to build_docs.py.
This forces the hugo server to bind the http server a port on the global
interface (as opposed to just on localhost, which is the default).

For other use unforeseen use-cases this also adds a `--hugo-opts` option to pass additional arguments directly to hugo.

This is important when using hugo on build farms with older (or no) browser
support, and no support for ssh tunneling.

Signed-off-by: Martin Lueker-Boden <martin.lueker-boden@wdc.com>

martin-lueker commented:
Thanks @asb for suggesting this!

rswarbrick commented on util/build_docs.py:
```diff
@@ -404,7 +404,7 @@ def install_hugo(install_dir):
 
 
-def invoke_hugo(preview, hugo_bin_path):
+def invoke_hugo(preview, hugo_bin_path, bind_wan):
```
Nit: Maybe it would be better to switch around the first or last two arguments? I think that `preview` and `bind_wan` are similar to each other, and different from `hugo_bin_path`. So I think this should be:
```python3
 def invoke_hugo(preview, bind_wan, hugo_bin_path):
```
or the other way around.

martin-lueker replied:
Hi @rswarbrick, I added one more arg in the last commit, but then I put them altogether and put the bin path last.

rswarbrick reviewed (approved):
Nitty comment about argument order, but this looks good to me.

towoe commented:
Thanks Martin. I am not sure what the use case is for you. Will you run the web server on a remote server and want to browse the documentation from your machine? Does this work for you? I tried this at my side but there are some problems, first, redirects with aliases are not working as they still point to localhost and second, the reference to 'style.css' is also done with localhost, so it is not a pleasant experience. But the text is there, so for terminal based browsing it might be okay ;) But it might be just a problem with my setup...

asb commented:
I think the other implementation approach that's worthy of consideration is to provide some kind of "--hugo-args` option (or an environment variable) to allow extra hugo arguments to be passed in a flexible way, on the basis that there might be argument beyond this one that are helpful. This use case could be listed as an example in that documentation.

Does anyone have strong feelings one way or another? Per @towoe's comment, I wondered if you need to have `--baseURL` set to something appropriate too for this to work as expected?

martin-lueker commented:
> Thanks Martin. I am not sure what the use case is for you. Will you run the web server on a remote server and want to browse the documentation from your machine? Does this work for you? I tried this at my side but there are some problems, first, redirects with aliases are not working as they still point to localhost and second, the reference to 'style.css' is also done with localhost, so it is not a pleasant experience. But the text is there, so for terminal based browsing it might be okay ;) But it might be just a problem with my setup...

Hi @towoe, 
  Interesting.  Maybe there is some subtlety I'm missing, but thus far it looks pretty good on my end.  If you have time it would be great if you could send a screenshot to illustrate your problems.  However maybe @asb is right that there are some more arguments I could add to straighten this out.
  As far as use case, you have it right.  We are *always* developing on a remote compute farm, because that is where our tool licenses are for our proprietary machines (and frankly, that is just how our IT folks want to support most us)  There are some options for viewing, such as:
- Launching a remote browser, which is clumsy at best, especially since the browsers are old. Worse still though, this is not working for some users for some system reason I don't understand.
- ssh tunnels to connect the "browser localhost" to the "server localhost".  This works on many Linux systems, but direct ssh is disabled on our compute farm.  So this option requires us to have a split workflow (proprietary tools on one host, documentation on another), which is suboptimal.
- Linux VM's just for browsing doc-edits, also suboptimal.
- The solution outlined here.

As an aside, this is actually the last hiccup to having a completely working OpenTitan flow on our RHEL6 compute farm servers.  Many of the other fixes required doing strange things like creating custom toolchains (both native and RISCV) or using using [patchelf](https://github.com/NixOS/patchelf), to point newer binaries (like hugo-extended) to a more modern version of glibc.

martin-lueker commented:
> I think the other implementation approach that's worthy of consideration is to provide some kind of "--hugo-args` option (or an environment variable) to allow extra hugo arguments to be passed in a flexible way, on the basis that there might be argument beyond this one that are helpful. This use case could be listed as an example in that documentation.
> 
> Does anyone have strong feelings one way or another? Per @towoe's comment, I wondered if you need to have `--baseURL` set to something appropriate too for this to work as expected?

Hi @asb, this is actually a great idea.  I could actually continue to tweak our setup without causing to much churn on the main repo.  Thanks, I'll try that.

imphil commented:
Btw, you should be able to commpile Hugo yourself on RHEL6 and use it there; it's just the pre-compiled version of Hugo that doesn't work.

martin-lueker commented:
> I think the other implementation approach that's worthy of consideration is to provide some kind of "--hugo-args` option (or an environment variable) to allow extra hugo arguments to be passed in a flexible way, on the basis that there might be argument beyond this one that are helpful. This use case could be listed as an example in that documentation.
> 
> Does anyone have strong feelings one way or another? Per @towoe's comment, I wondered if you need to have `--baseURL` set to something appropriate too for this to work as expected?

As far as args vs. environment variables, environment variables are super convenient, but I often worry about environment contamination causing trouble for other users who aren't aware of them.   I might actually do something like this: make a variable `HUGO_ARGS`, but then ignore it unless there is a flag `--env-hugo-args`.   Any thoughts on that? 

If that is too baroque, then a simple --hugo-args (and a wdc-side script that adds the right args) might be fine.

martin-lueker commented:
> Btw, you should be able to compile Hugo yourself on RHEL6 and use it there; it's just the pre-compiled version of Hugo that doesn't work.

Yeah, but compiling was a problem on it's own for the tool and dependency setup.  It might get easier as we make progress with our native toolchain.  In the meantime patchelf was far far easier (though I did have to use the right version so maybe that is not the best long term solution).  I'll probably be writing a footnote doc to link to the software page.

imphil commented:
I just gave it a try and compiled hugo on RHEL6, you can download the binary here: https://drive.google.com/file/d/1eLGchH-aX52M2pxuGFmSt6aeGZMPLxzN/view?usp=sharing

If that works for you we can modify the build_docs.py script to use this build by default when running on RHEL6, providing users on RHEL6 the same experience others.

martin-lueker commented:
> I just gave it a try and compiled hugo on RHEL6, you can download the binary here: https://drive.google.com/file/d/1eLGchH-aX52M2pxuGFmSt6aeGZMPLxzN/view?usp=sharing
> 
> If that works for you we can modify the build_docs.py script to use this build by default when running on RHEL6, providing users on RHEL6 the same experience others.

Thanks @imphil!  I can give it a try.

martin-lueker commented:
@towoe: Interestingly enough I now see what you are talking about.  (I seem to have some hysteresis in my page views, presumably due to browser caches?)   Though @asb, your `baseURL` suggestion seems to work very well.   I added `baseURL` argument into the settings added by the `--bind-wan` option.

@towoe, could you try it now?

As for the more generic approach (`--hugo-opts` or environment variables), I added that too, and that should help for others wanting to add additional options.   However for the simple case, the `--bind-wan` option is a bit easier to use as it requires less scripting on the part of the user (have to look up the hostname etc.) and it would be nice to not need to make another script to wrap this script.   So I've left them in.

I played with the option of environment variables, however that is not quite compatible with our environment management system.  On this compute farm we are often jumping from node to node, and because of some peculiarities of the environment management tools on our system keeping the "baseURL" argument in an environment variable could cause confusion.

That said, I'd like to add both the `--hugo-opts` and `--bind-wan` flags.

martin-lueker commented:
BTW: Thanks for the review @swarbrick.   Does anyone else have any thoughts or reservations on merging this as is with these two new options?  I ask because we are having active conversations right now about our remaining capability gaps on the WDC side servers, and this is one of the few remaining items. 
(BTW it is really exciting to be able to say that, I really appreciate the help from everyone who helped us get this far!  I think @imphil in particular deserves a big shout out, for all his support builds of various tools.)

imphil commented:
LGTM if it does the trick for you!

About the hugo binary: did the binary I linked to work for you? And if so, do you want us to distribute it somewhere for you to install locally, or do you have ways to produce a working hugo binary for RHEL6 yourself (and make it available e.g. through `module load`)?

martin-lueker commented:
Thanks @imphil!  You bring up a good point though, I'll try to get at least one more person to try this on our system (to make sure I'm not missing something else).

martin-lueker commented:
Thanks all.  I've had a couple people test the latest version from our side, and aside from a minor bug (`--bind-wan` without `--preview` crashed... fixed) it seems to work well. Merging..
